File: vscode/description_test.go
package vscode

import "testing"

func int64Ptr(v int64) *int64 { return &v }

// TestRenderDescriptionFixedClassOrderAndSeparator holds keel/ac-554: an item
// carrying producer prose, a typed last run, typed desired-state facts and a
// warning-severity finding renders those four classes in the order the
// renderer declares, joined by the one separator it declares — and reordering
// the fields of the received item changes nothing, because the renderer reads
// fields by name rather than by arrival.
//
// DHF-TEST: keel/requirement-139
func TestRenderDescriptionFixedClassOrderAndSeparator(t *testing.T) {
	item := TestItem{
		ID:          "keel::lane::a",
		Description: "the lane prose",
		LastRun:     &LastRunFacts{DurationMS: int64Ptr(9800)},
		DesiredStateRow: &DesiredStateRowFacts{
			Current: "empty",
			Action:  "reconcile",
			Active:  true,
		},
		Findings: []Finding{{Rule: "lane-order", Severity: "warning", Message: "order drifted"}},
	}

	got := RenderDescription(item, DefaultDisplayConfig())
	want := "the lane prose; · last 9.8s; current=empty; action=reconcile; active=true; lane-order warning: order drifted"
	if got != want {
		t.Fatalf("RenderDescription()\n got: %q\nwant: %q", got, want)
	}
}

// TestRenderDescriptionToggleSuppressesExactlyItsClass holds keel/ac-555:
// clearing exactly one display toggle removes that class from the rendered
// string and leaves every other class byte-identical to the all-enabled
// render.
//
// DHF-TEST: keel/requirement-139
func TestRenderDescriptionToggleSuppressesExactlyItsClass(t *testing.T) {
	item := TestItem{
		ID:              "keel::lane::a",
		Description:     "the lane prose",
		LastRun:         &LastRunFacts{DurationMS: int64Ptr(9800)},
		DesiredStateRow: &DesiredStateRowFacts{Current: "empty", Action: "reconcile", Active: true},
		Findings:        []Finding{{Rule: "lane-order", Severity: "warning", Message: "order drifted"}},
	}

	for _, tc := range []struct {
		name    string
		display DisplayConfig
		want    string
	}{
		{
			name:    "description off",
			display: DisplayConfig{LastRun: true, DesiredState: true, Findings: true},
			want:    "· last 9.8s; current=empty; action=reconcile; active=true; lane-order warning: order drifted",
		},
		{
			name:    "last run off",
			display: DisplayConfig{Description: true, DesiredState: true, Findings: true},
			want:    "the lane prose; current=empty; action=reconcile; active=true; lane-order warning: order drifted",
		},
		{
			name:    "desired state off",
			display: DisplayConfig{Description: true, LastRun: true, Findings: true},
			want:    "the lane prose; · last 9.8s; lane-order warning: order drifted",
		},
		{
			name:    "findings off",
			display: DisplayConfig{Description: true, LastRun: true, DesiredState: true},
			want:    "the lane prose; · last 9.8s; current=empty; action=reconcile; active=true",
		},
		{
			name:    "every class off",
			display: DisplayConfig{},
			want:    "",
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if got := RenderDescription(item, tc.display); got != tc.want {
				t.Fatalf("RenderDescription()\n got: %q\nwant: %q", got, tc.want)
			}
		})
	}
}

// TestRenderDescriptionIsProducerIndependent holds keel/ac-553: two discovery
// items carrying byte-identical typed facts and prose render a byte-identical
// description, whatever else distinguishes the producers that sent them.
//
// DHF-TEST: keel/requirement-139
func TestRenderDescriptionIsProducerIndependent(t *testing.T) {
	facts := func(id string, limitations []string) TestItem {
		return TestItem{
			ID:              id,
			Label:           "shared label",
			Limitations:     limitations,
			Description:     "the same prose",
			LastRun:         &LastRunFacts{DurationMS: int64Ptr(1500)},
			DesiredStateRow: &DesiredStateRowFacts{Current: "running", Action: "reuse", Active: true},
			Findings:        []Finding{{Rule: "r", Severity: "error", Message: "m"}},
		}
	}

	fromKeelDev := RenderDescription(facts("keel::lane::a", []string{"the same prose", "· last 1.5s"}), DefaultDisplayConfig())
	fromOpenbrainDev := RenderDescription(facts("openbrain::lane::a", nil), DefaultDisplayConfig())
	if fromKeelDev != fromOpenbrainDev {
		t.Fatalf("producers rendered differently\n keel-dev: %q\n openbrain-dev: %q", fromKeelDev, fromOpenbrainDev)
	}
	if fromKeelDev == "" {
		t.Fatal("RenderDescription() rendered nothing for an item carrying every fact class")
	}
}

// TestFormatLastRunOmitsAnAbsentMeasurement holds the absence half of
// keel/ac-554: a missing measurement contributes no segment at all rather than
// a zero-valued one, so the separator never leads or doubles.
//
// DHF-TEST: keel/requirement-139
func TestFormatLastRunOmitsAnAbsentMeasurement(t *testing.T) {
	for _, tc := range []struct {
		name string
		last *LastRunFacts
		want string
	}{
		{name: "no last run at all", last: nil, want: ""},
		{name: "run without a measured duration", last: &LastRunFacts{}, want: ""},
		{name: "negative duration", last: &LastRunFacts{DurationMS: int64Ptr(-1)}, want: ""},
		{name: "measured zero", last: &LastRunFacts{DurationMS: int64Ptr(0)}, want: "· last 0.0s"},
		{name: "sub-minute", last: &LastRunFacts{DurationMS: int64Ptr(9800)}, want: "· last 9.8s"},
		{name: "at the ninety-second boundary", last: &LastRunFacts{DurationMS: int64Ptr(90000)}, want: "· last 90.0s"},
		{name: "over the boundary", last: &LastRunFacts{DurationMS: int64Ptr(192000)}, want: "· last 3m 12s"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if got := FormatLastRun(tc.last); got != tc.want {
				t.Fatalf("FormatLastRun()\n got: %q\nwant: %q", got, tc.want)
			}
		})
	}
}

// TestRenderDescriptionOfAnItemWithoutFactsIsEmpty holds the consumer-side
// fallback precondition: an item carrying no typed fact renders nothing, so a
// consumer can tell "the renderer had nothing to say" from "the renderer said
// something short" without inspecting the item itself.
//
// DHF-TEST: keel/requirement-139
func TestRenderDescriptionOfAnItemWithoutFactsIsEmpty(t *testing.T) {
	item := TestItem{ID: "openbrain::lane::a", Limitations: []string{"current=empty", "action=reconcile"}}
	if got := RenderDescription(item, DefaultDisplayConfig()); got != "" {
		t.Fatalf("RenderDescription() of an item carrying no facts = %q, want empty", got)
	}
	if HasRenderableFacts(item) {
		t.Fatal("HasRenderableFacts() reported facts on an item carrying only limitations")
	}
}

File: vsix/src/test/suite/description.test.ts
import * as assert from 'node:assert/strict';
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import * as vscode from 'vscode';
import {
  composeDescription,
  defaultDisplayConfig,
  descriptionSeparator,
  displayClassOrder,
  hasRenderableFacts,
  parseDisplayConfig,
  renderDescription,
  DisplayConfig
} from '../../description';
import { adapterConfig, configRelativePath, currentConfigVersion, defaultConfigTemplate, readAdapterConfig } from '../../bridgeAdapter';
import { DiscoveryItem } from '../../protocol';
import { publishDiscovery } from '../../tree';

interface GoldenCase {
  name: string;
  item: DiscoveryItem;
  display: DisplayConfig;
  expected: string;
  has_facts: boolean;
}

interface GoldenFixture {
  version: number;
  separator: string;
  class_order: string[];
  defaults: { display: DisplayConfig };
  cases: GoldenCase[];
}

function readGoldenFixture(): GoldenFixture {
  const fixturePath = path.resolve(__dirname, '../../../../testdata/description-golden.json');
  return JSON.parse(fs.readFileSync(fixturePath, 'utf8')) as GoldenFixture;
}

function makeWorkspace(prefix: string, config: unknown): string {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), prefix));
  fs.mkdirSync(path.join(root, '.vscode'), { recursive: true });
  fs.writeFileSync(path.join(root, configRelativePath), JSON.stringify(config, null, 2) + '\n');
  return root;
}

suite('description composition', () => {
  // The VSIX half of keel/ac-561. The Go renderer asserts the same committed
  // file, so changing either implementation alone reds a gate.
  // DHF-TEST: keel/requirement-139
  test('req-139 the composer renders every golden fixture case', () => {
    const fixture = readGoldenFixture();
    assert.equal(fixture.version, 1);
    assert.ok(fixture.cases.length > 0, 'golden fixture carries no cases');
    for (const testCase of fixture.cases) {
      assert.equal(renderDescription(testCase.item, testCase.display), testCase.expected, testCase.name);
      assert.equal(hasRenderableFacts(testCase.item), testCase.has_facts, testCase.name);
    }
  });

  // DHF-TEST: keel/requirement-139
  test('req-139 the composer declares the same contract as the fixture', () => {
    const fixture = readGoldenFixture();
    assert.equal(fixture.separator, descriptionSeparator);
    assert.deepEqual(fixture.class_order, displayClassOrder);
    assert.deepEqual(fixture.defaults.display, defaultDisplayConfig());
  });

  // keel/ac-554: field order inside the received object cannot reach the output.
  // DHF-TEST: keel/requirement-139
  test('req-139 reordering the received JSON object changes nothing', () => {
    const forward = JSON.parse(JSON.stringify({
      id: 'keel::lane::a',
      label: 'A',
      kind: 'lane',
      runnable: true,
      profiles: [],
      description: 'prose',
      last_run: { at: '2026-08-18T10:00:00Z', duration_ms: 9800 },
      findings: [{ rule: 'r', severity: 'warning', message: 'm' }]
    })) as DiscoveryItem;
    const reversed = JSON.parse(JSON.stringify({
      findings: [{ rule: 'r', severity: 'warning', message: 'm' }],
      last_run: { at: '2026-08-18T10:00:00Z', duration_ms: 9800 },
      description: 'prose',
      profiles: [],
      runnable: true,
      kind: 'lane',
      label: 'A',
      id: 'keel::lane::a'
    })) as DiscoveryItem;
    assert.equal(
      renderDescription(forward, defaultDisplayConfig()),
      renderDescription(reversed, defaultDisplayConfig())
    );
    assert.equal(renderDescription(forward, defaultDisplayConfig()), 'prose; · last 9.8s; r warning: m');
  });

  // keel/ac-553: the rendered text depends on the facts, never on the producer.
  // DHF-TEST: keel/requirement-139
  test('req-139 two producers emitting identical facts render identical text', () => {
    const facts = (id: string): DiscoveryItem => ({
      id,
      label: 'shared',
      kind: 'lane',
      runnable: true,
      profiles: [],
      description: 'the same prose',
      last_run: { at: '2026-08-18T10:00:00Z', duration_ms: 1500 },
      desired_state_row: { current: 'running', action: 'reuse', active: true },
      findings: [{ rule: 'r', severity: 'error', message: 'm' }]
    });
    const fromKeelDev = composeDescription(facts('keel::lane::a'), defaultDisplayConfig());
    const fromOpenbrainDev = composeDescription(facts('openbrain::lane::a'), defaultDisplayConfig());
    assert.equal(fromKeelDev, fromOpenbrainDev);
    assert.ok(fromKeelDev && fromKeelDev.length > 0);
  });

  // The fallback is retired: with the prose array gone from the wire, an item
  // carrying no typed fact renders nothing whatever producer emitted it. This is
  // asserted per producer shape so a producer that silently stopped emitting
  // typed facts is an observed fact rather than an invisible default.
  // DHF-TEST: keel/requirement-138, keel/requirement-139
  test('req-138 an openbrain-dev item with no typed facts renders nothing, the fallback being retired', () => {
    const item: DiscoveryItem = {
      id: 'openbrain::lane::a',
      label: 'A',
      kind: 'lane',
      runnable: true,
      profiles: []
    };
    assert.equal(hasRenderableFacts(item), false);
    assert.equal(composeDescription(item, defaultDisplayConfig()), undefined);
  });

  // DHF-TEST: keel/requirement-139
  test('req-139 a keel-demo-dev item with no typed facts renders nothing at all', () => {
    const item: DiscoveryItem = { id: 'demo::lane::a', label: 'A', kind: 'lane', runnable: true, profiles: [] };
    assert.equal(hasRenderableFacts(item), false);
    assert.equal(composeDescription(item, defaultDisplayConfig()), undefined);
  });

  // A suppressed class must stay suppressed: there is no prose channel left to
  // resurrect what the toggles removed.
  // DHF-TEST: keel/requirement-138, keel/requirement-139
  test('req-139 a typed item with every class disabled renders nothing', () => {
    const item: DiscoveryItem = {
      id: 'keel::lane::a',
      label: 'A',
      kind: 'lane',
      runnable: true,
      profiles: [],
      description: 'the lane prose',
      last_run: { at: '2026-08-18T10:00:00Z', duration_ms: 9800 }
    };
    assert.equal(composeDescription(item, { description: false, lastRun: false, desiredState: false, findings: false, ordinal: false }), undefined);
  });

  // keel/ac-555 end to end: the published tree carries the composed text, and
  // clearing one toggle removes exactly that class from it.
  // DHF-TEST: keel/requirement-139
  test('req-139 the published tree renders the composed description', () => {
    const controller = vscode.tests.createTestController('keel-description-test', 'Keel description test');
    try {
      const item: DiscoveryItem = {
        id: 'keel::lane::a',
        label: 'A',
        kind: 'lane',
        runnable: true,
        profiles: [],
        description: 'the lane prose',
        last_run: { at: '2026-08-18T10:00:00Z', duration_ms: 9800 },
        findings: [{ rule: 'lane-order', severity: 'warning', message: 'order drifted' }]
      };
      const discovery = { version: 1 as const, workspace: '/tmp', generated_at: '2026-08-18T10:00:00Z', items: [item] };

      const published = publishDiscovery(controller, '/tmp', discovery, 0, defaultDisplayConfig());
      assert.equal(published.itemsById.get('keel::lane::a')?.description, 'the lane prose; · last 9.8s; lane-order warning: order drifted');

      const withoutLastRun = publishDiscovery(controller, '/tmp', discovery, 0, { description: true, lastRun: false, desiredState: true, findings: true, ordinal: false });
      assert.equal(withoutLastRun.itemsById.get('keel::lane::a')?.description, 'the lane prose; lane-order warning: order drifted');
    } finally {
      controller.dispose();
    }
  });
});

suite('display configuration', () => {
  // keel/ac-566.
  // DHF-TEST: keel/requirement-139
  test('req-139 an unknown key inside the display block is rejected at config read', () => {
    const root = makeWorkspace('keel-display-unknown-', {
      version: currentConfigVersion,
      command: 'bin/keel-dev',
      args: [],
      displayName: 'Keel',
      display: { finding: false }
    });
    assert.throws(() => readAdapterConfig(root), /finding/);
  });

  // DHF-TEST: keel/requirement-139
  test('req-139 a non-boolean toggle is rejected at config read', () => {
    const root = makeWorkspace('keel-display-nonbool-', {
      version: currentConfigVersion,
      command: 'bin/keel-dev',
      args: [],
      displayName: 'Keel',
      display: { findings: 'no' }
    });
    assert.throws(() => readAdapterConfig(root), /findings/);
  });

  // DHF-TEST: keel/requirement-139
  test('req-139 an absent display block and an absent key both mean enabled', () => {
    const absent = makeWorkspace('keel-display-absent-', {
      version: currentConfigVersion,
      command: 'bin/keel-dev',
      args: [],
      displayName: 'Keel'
    });
    assert.deepEqual(adapterConfig(absent).display, defaultDisplayConfig());

    const partial = makeWorkspace('keel-display-partial-', {
      version: currentConfigVersion,
      command: 'bin/keel-dev',
      args: [],
      displayName: 'Keel',
      display: { findings: false }
    });
    assert.deepEqual(adapterConfig(partial).display, { description: true, lastRun: true, desiredState: true, findings: false, ordinal: false });
  });

  // DHF-TEST: keel/requirement-139
  test('req-139 parseDisplayConfig names the unknown key it refuses', () => {
    assert.throws(() => parseDisplayConfig({ lastrun: true }), /lastrun/);
    assert.deepEqual(parseDisplayConfig(undefined), defaultDisplayConfig());
    assert.deepEqual(parseDisplayConfig({}), defaultDisplayConfig());
  });

  // DHF-TEST: keel/requirement-59, keel/requirement-139
  test('req-139 the default template states every class enabled at the current version', () => {
    const parsed = JSON.parse(defaultConfigTemplate()) as { version: number; display: DisplayConfig };
    assert.equal(parsed.version, currentConfigVersion);
    assert.deepEqual(parsed.display, defaultDisplayConfig());
  });
});
